### PR TITLE
Add g_usb_device_get_created() to better detect emulated replug

### PR DESCRIPTION
**Pull request by hughsie**

### CODE REVIEW

File: gusb/gusb-device.c [hunk 1]
```diff
@@ -239,6 +239,22 @@ _g_usb_device_load(GUsbDevice *self, JsonObject *json_object, GError **error)
 		g_free(priv->platform_id);
 		priv->platform_id = g_strdup(tmp);
 	}
+	tmp = json_object_get_string_member_with_default(json_object, "Created", NULL);
+	if (tmp != NULL) {
+		g_autoptr(GDateTime) created_new = g_date_time_new_from_iso8601 (tmp, NULL);
```
Comment (nacho) [inline gusb/gusb-device.c:244]: no space before (?

File: gusb/gusb-device.c [hunk 2]
```diff
@@ -239,6 +239,22 @@ _g_usb_device_load(GUsbDevice *self, JsonObject *json_object, GError **error)
...
+		if (created_new == NULL) {
+			g_set_error(error,
+				    G_IO_ERROR,
+				    G_IO_ERROR_INVALID_DATA,
+				    "cannot parse ISO8601 date: %s",
```
Comment (nacho) [inline gusb/gusb-device.c:249]: Cannot?

File: gusb/gusb-device.c [hunk 3]
```diff
@@ -450,6 +466,27 @@ _g_usb_device_save(GUsbDevice *self, JsonBuilder *json_builder, GError **error)
...
+ *
+ * This can be used as an indicator if the device replugged, as the vendor and product IDs may not
+ * change for some devices. Use `g_date_time_equal()` to verify equality.
+ *
+ * Returns: (transfer none): a #GDateTime
+ *
+ * Since: 0.4.5
+ **/
+GDateTime *
+g_usb_device_get_created(GUsbDevice *self)
```
Comment (nacho) [inline gusb/gusb-device.c:483]: get_created_date_time?

File: gusb/gusb-device.c (hunk 1 above)
Comment (hughsie) [reply gusb/gusb-device.c:244]: Fixed, thanks.

File: gusb/gusb-device.c (hunk 2 above)
Comment (hughsie) [reply gusb/gusb-device.c:249]: Fixed, thanks.

File: gusb/gusb-device.c (hunk 3 above)
Comment (hughsie) [reply gusb/gusb-device.c:483]: On the logic that we might have `g_usb_device_get_created_iso8601()` in the future?

Comment (nacho) [reply gusb/gusb-device.c:483]: actually dunno, maybe just get_created is enough since from the datetime you can get the iso format

**Review by nacho — APPROVED** [review]

**Comment by hughsie** [comment]
@fdanis-oss I think this might solve a few of our problems.

**Comment by hughsie** [comment]
> should this be use to invalidate the device probe in fu_usb_device_set_dev

Ohh! Maybe --  I was thinking of doing something like:

	diff --git a/src/fu-usb-backend.c b/src/fu-usb-backend.c
	index 194345883..ae7303945 100644
	--- a/src/fu-usb-backend.c
	+++ b/src/fu-usb-backend.c
	@@ -56,10 +56,14 @@ fu_usb_backend_device_added_cb(GUsbContext *ctx, GUsbDevice *usb_device, FuBacke
		/* is emulated? */
		device_tmp = fu_backend_lookup_by_id(backend, g_usb_device_get_platform_id(usb_device));
		if (device_tmp != NULL && fu_device_has_flag(device_tmp, FWUPD_DEVICE_FLAG_EMULATED)) {
	-               if (g_usb_device_get_vid(usb_device) ==
	-                       fu_usb_device_get_vid(FU_USB_DEVICE(device_tmp)) &&
	-                   g_usb_device_get_pid(usb_device) ==
	-                       fu_usb_device_get_pid(FU_USB_DEVICE(device_tmp))) {
	+               GUsbDevice *usb_device_tmp = fu_usb_device_get_dev(FU_USB_DEVICE(device_tmp));
	+#if G_USB_CHECK_VERSION(0, 4, 5)
	+               if (g_date_time_equal(g_usb_device_get_created(usb_device),
	+                                     g_usb_device_get_created(usb_device_tmp))) {
	+#else
	+               if (g_usb_device_get_vid(usb_device) == g_usb_device_get_vid(usb_device_tmp) &&
	+                   g_usb_device_get_pid(usb_device) == g_usb_device_get_pid(usb_device_tmp)) {
	+#endif
				g_debug("replacing GUsbDevice of emulated device %s",
					fu_usb_device_get_platform_id(FU_USB_DEVICE(device_tmp)));
				fu_usb_device_set_dev(FU_USB_DEVICE(device_tmp), usb_device);

but I'm just playing around now. My gut instinct is that we really do want to simulate the replug and not do something like https://github.com/fwupd/fwupd/pull/5478 -- as the "failed to submit control transfer because the device went away" is actually a really tricky thing to get right, and I think it makes sense to emulate that. If it all gets too crazy we can do it your simple way.

**Comment by fdanis-oss** [comment]
yes, it works fine with your change

**Review by fdanis-oss — APPROVED** [review]

**Pull request merged**